keanu-project/src/test/java/io/improbable/keanu/network/NetworkSnapshotTest.java

```diff
@@ -0,0 +1,115 @@
+    }
+
+    @Test
+    public void itSetsTheStateOfAVertex() {
+        VertexState<Object> s1 = VertexState.nullState();
+        VertexState<Object> s2 = VertexState.nullState();
+        Vertex<Object> v1 = mock(Vertex.class);
+        Vertex<Object> v2 = mock(Vertex.class);
+        when(v1.getState()).thenReturn(s1);
+        when(v2.getState()).thenReturn(s2);
+        doCallRealMethod().when(v1).setState(any(VariableState.class));
+        doCallRealMethod().when(v2).setState(any(VariableState.class));
+
+        NetworkSnapshot snapshot = NetworkSnapshot.create(ImmutableSet.of(v1, v2));
+        verify(v1).getState();
+        verify(v2).getState();
+
+        snapshot.apply();
+        verify(v1).setState(s1);
+        verify(v2).setState(s2);
+    }
+
+    @Test
+    public void itRestoresTheValueOfAnUnobservedVertex() {
+        DoubleTensor originalValue = DoubleTensor.create(1., 2., 3.);
+        DoubleTensor otherValue = DoubleTensor.create(4., 5., 6.);
+        Vertex vertex = new GaussianVertex(1., 0.);
+        vertex.setValue(originalValue);
+        NetworkSnapshot snapshot = NetworkSnapshot.create(ImmutableSet.of(vertex));
+
+        vertex.setValue(otherValue);
+        assertThat(vertex.getValue(), equalTo(otherValue));
+        assertThat(vertex.isObserved(), is(false));
+
+        snapshot.apply();
+        assertThat(vertex.getValue(), equalTo(originalValue));
+        assertThat(vertex.isObserved(), is(false));
+    }
+
+    @Test
+    public void itRestoresTheValueOfAnObservedVertex() {
+        DoubleTensor originalValue = DoubleTensor.create(1., 2., 3.);
+        DoubleTensor otherValue = DoubleTensor.create(4., 5., 6.);
+        Vertex vertex = new GaussianVertex(1., 0.);
+
+        vertex.observe(originalValue);
+        assertThat(vertex.getValue(), equalTo(originalValue));
+        assertThat(vertex.isObserved(), is(true));
+
+        NetworkSnapshot snapshot = NetworkSnapshot.create(ImmutableSet.of(vertex));
+
+        vertex.observe(otherValue);
+        assertThat(vertex.getValue(), equalTo(otherValue));
+        assertThat(vertex.isObserved(), is(true));
+
+        snapshot.apply();
+        assertThat(vertex.getValue(), equalTo(originalValue));
+        assertThat(vertex.isObserved(), is(true));
+    }
+
+    @Test
+    public void itRestoresTheObservedStatusOfAnObservedVertex() {
+        DoubleTensor originalValue = DoubleTensor.create(1., 2., 3.);
+        DoubleTensor otherValue = DoubleTensor.create(4., 5., 6.);
+        Vertex vertex = new GaussianVertex(1., 0.);
+
+        vertex.observe(originalValue);
+        assertThat(vertex.getValue(), equalTo(originalValue));
+        assertThat(vertex.isObserved(), is(true));
+
+        NetworkSnapshot snapshot = NetworkSnapshot.create(ImmutableSet.of(vertex));
+
+        vertex.unobserve();
+        assertThat(vertex.getValue(), equalTo(originalValue));
+        assertThat(vertex.isObserved(), is(false));
+
+        vertex.setValue(otherValue);
```

Comment:
nit: you're already testing the changing of the value in the previous 2 tests. You only need the check on `isObserved` here